xds/internal/xdsclient/watchers_test.go

```diff
@@ -0,0 +1,599 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xdsclient
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/xds/internal/xdsclient/bootstrap"
+	"google.golang.org/grpc/xds/internal/xdsclient/pubsub"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource/version"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+// testWatchSetup starts the client and makes a watch on it for the resourceName.
+// The caller can make new watchers, or start sending updates for this watch
+// afterwards.
+//
+// Note that this function checks for the controller after watching. This is the
+// new behavior after federation.
+func testWatchSetup(ctx context.Context, t *testing.T, typ xdsresource.ResourceType, resourceName string, overrideWatchExpiryTimeout bool) (_ *clientImpl, _ *testController, _ pubsub.UpdateHandler, updateCh *testutils.Channel, cancelWatch func()) {
```

Comment:
I feel like we are sacrificing test code readability for test code brevity. This function does a lot of things, and not all callers use all the return values. Do you think we can simplify things at the code of writing repetitive test code?

See: https://engdoc.corp.google.com/eng/doc/tott/episodes/598.md?cl=head

Reply:
Done.

Federation made this util function more complex, because now controller is only available after a watch.

But the code doesn't look that cleaner even after splitting... There are just too many steps.